File: oil_distribution/dashboard_reports/report/reserved_stock/reserved_stock.py
import frappe
from frappe.utils import flt
from collections import defaultdict


def execute(filters=None):
	if not filters:
		filters = {}

	fields = [
		"sr.name as reservation",
		"sr.company",
		"sr.warehouse",
		"sr.item",
		"sr.reserved_qty",
		"sr.reserved_for",
		"sr.posting_date",
		"sr.status",
		"sr.sales_order",
		"sr.stock_entry",
	]

	conditions = ""
	if filters.get("company"):
		conditions += " and sr.company = %(company)s"
	if filters.get("warehouse"):
		conditions += " and sr.warehouse = %(warehouse)s"
	if filters.get("item"):
		conditions += " and sr.item = %(item)s"
	if filters.get("status"):
		conditions += " and sr.status = %(status)s"

	query = """
		select {fields}
		from `tabStock Reservation` sr
		where sr.docstatus = 1 {conditions}
		order by sr.posting_date desc
	""".format(
		fields=", ".join(fields),
		conditions=conditions,
	)

	data = frappe.db.sql(query, filters, as_dict=True)

	columns = [
		{"label": "Reservation", "fieldname": "reservation", "fieldtype": "Link", "options": "Stock Reservation", "width": 180},
		{"label": "Company", "fieldname": "company", "fieldtype": "Link", "options": "Company", "width": 180},
		{"label": "Warehouse", "fieldname": "warehouse", "fieldtype": "Link", "options": "Warehouse", "width": 180},
		{"label": "Item", "fieldname": "item", "fieldtype": "Link", "options": "Item", "width": 150},
		{"label": "Reserved Qty", "fieldname": "reserved_qty", "fieldtype": "Float", "width": 120},
		{"label": "Reserved For", "fieldname": "reserved_for", "width": 150},
		{"label": "Date", "fieldname": "posting_date", "fieldtype": "Date", "width": 120},
		{"label": "Status", "fieldname": "status", "width": 120},
		{"label": "Sales Order", "fieldname": "sales_order", "fieldtype": "Link", "options": "Sales Order", "width": 150},
		{"label": "Stock Entry", "fieldname": "stock_entry", "fieldtype": "Link", "options": "Stock Entry", "width": 150},
	]

	# Chart: Reserved qty by company (bar)
	company_reserved = defaultdict(float)
	for row in data:
		company_reserved[row.company] += flt(row.reserved_qty)

	chart = None
	if company_reserved:
		chart = {
			"data": {
				"labels": list(company_reserved.keys()),
				"datasets": [{"name": "Reserved Qty", "values": list(company_reserved.values())}],
			},
			"type": "bar",
		}

	# Pie chart for status distribution
	status_counts = defaultdict(int)
	for row in data:
		status_counts[row.status or "Reserved"] += 1

	pie_chart = None
	if status_counts:
		pie_chart = {
			"data": {
				"labels": list(status_counts.keys()),
				"datasets": [{"name": "Reservations", "values": list(status_counts.values())}],
			},
			"type": "pie",
		}

	# Report summary
	total_reservations = len(data)
	total_qty = sum(flt(r.reserved_qty) for r in data)
	active = sum(1 for r in data if r.status == "Reserved")
	released = sum(1 for r in data if r.status == "Released")

	report_summary = [
		{"label": "Total Reservations", "value": total_reservations, "indicator": "blue"},
		{"label": "Total Reserved Qty", "value": total_qty, "indicator": "orange"},
		{"label": "Active", "value": active, "indicator": "green"},
		{"label": "Released", "value": released, "indicator": "red"},
	]

	return columns, data, None, chart, report_summary

File: oil_distribution/oil_distribution/report/reserved_stock/reserved_stock.js
frappe.query_reports["Reserved Stock"] = {
	"filters": [
		{ "fieldname": "company", "label": "Company", "fieldtype": "Link", "options": "Company", "default": frappe.defaults.get_default("company") },
		{ "fieldname": "warehouse", "label": "Warehouse", "fieldtype": "Link", "options": "Warehouse" },
		{ "fieldname": "item", "label": "Item", "fieldtype": "Link", "options": "Item" },
		{ "fieldname": "status", "label": "Status", "fieldtype": "Select", "options": "\nDraft\nReserved\nReleased\nCancelled" },
	],

	format: function (report) {
		var data = report.data || [];
		var summary = report.report_summary || [];

		var h = '<div style="padding:12px;font-family:-apple-system,BlinkMacSystemFont,Segoe UI,sans-serif;">';

		// KPI cards
		if (summary.length) {
			h += '<div style="display:grid;grid-template-columns:repeat(auto-fill,minmax(160px,1fr));gap:10px;margin-bottom:16px;">';
			summary.forEach(function (k) {
				var colors = { blue: '#3b82f6', green: '#10b981', orange: '#f59e0b', red: '#dc2626' };
				var bgColors = { blue: '#eff6ff', green: '#ecfdf5', orange: '#fef3c7', red: '#fef2f2' };
				var c = colors[k.indicator] || '#64748b';
				var bg = bgColors[k.indicator] || '#f8fafc';
				h += '<div style="background:' + bg + ';border:1px solid #f1f5f9;border-left:3px solid ' + c + ';border-radius:10px;padding:12px;">';
				h += '<div style="font-size:9px;font-weight:700;color:#94a3b8;text-transform:uppercase;">' + k.label + '</div>';
				h += '<div style="font-size:20px;font-weight:800;color:' + c + ';margin-top:2px;">' + k.value + '</div>';
				h += '</div>';
			});
			h += '</div>';
		}

		if (data.length) {
			// Group by status
			var byStatus = {};
			data.forEach(function (row) {
				var st = row.status || 'Reserved';
				if (!byStatus[st]) byStatus[st] = [];
				byStatus[st].push(row);
			});

			var statusColors = { 'Reserved': '#f59e0b', 'Draft': '#94a3b8', 'Released': '#10b981', 'Cancelled': '#dc2626' };

			Object.keys(byStatus).forEach(function (st) {
				var rows = byStatus[st];
				var sc = statusColors[st] || '#64748b';
				var totalQty = rows.reduce(function (s, r) { return s + (r.reserved_qty || 0); }, 0);

				h += '<div style="background:#fff;border:1px solid #e2e8f0;border-radius:12px;margin-bottom:12px;overflow:hidden;">';
				// Header
				h += '<div style="display:flex;align-items:center;justify-content:space-between;padding:12px 16px;border-bottom:1px solid #f1f5f9;background:#fafbfc;">';
				h += '<div style="display:flex;align-items:center;gap:8px;">';
				h += '<span style="width:10px;height:10px;border-radius:3px;background:' + sc + ';"></span>';
				h += '<span style="font-size:12px;font-weight:700;color:#1e293b;">' + st + '</span>';
				h += '<span style="font-size:9px;font-weight:700;padding:2px 8px;border-radius:10px;background:' + sc + '20;color:' + sc + ';">' + rows.length + ' reservations</span>';
				h += '</div>';
				h += '<span style="font-size:13px;font-weight:800;color:' + sc + ';">' + frappe.utils.fmt_money(totalQty) + ' L</span>';
				h += '</div>';

				// Table
				h += '<table style="width:100%;border-collapse:collapse;font-size:11px;">';
				h += '<thead><tr>';
				h += '<th style="text-align:left;padding:8px 12px;font-size:9px;font-weight:700;color:#94a3b8;text-transform:uppercase;border-bottom:1px solid #f1f5f9;">Reservation</th>';
				h += '<th style="text-align:left;padding:8px 12px;font-size:9px;font-weight:700;color:#94a3b8;text-transform:uppercase;border-bottom:1px solid #f1f5f9;">Company</th>';
				h += '<th style="text-align:left;padding:8px 12px;font-size:9px;font-weight:700;color:#94a3b8;text-transform:uppercase;border-bottom:1px solid #f1f5f9;">Warehouse</th>';
				h += '<th style="text-align:left;padding:8px 12px;font-size:9px;font-weight:700;color:#94a3b8;text-transform:uppercase;border-bottom:1px solid #f1f5f9;">Item</th>';
				h += '<th style="text-align:right;padding:8px 12px;font-size:9px;font-weight:700;color:#94a3b8;text-transform:uppercase;border-bottom:1px solid #f1f5f9;">Qty</th>';
				h += '<th style="text-align:left;padding:8px 12px;font-size:9px;font-weight:700;color:#94a3b8;text-transform:uppercase;border-bottom:1px solid #f1f5f9;">Reserved For</th>';
				h += '<th style="text-align:left;padding:8px 12px;font-size:9px;font-weight:700;color:#94a3b8;text-transform:uppercase;border-bottom:1px solid #f1f5f9;">Date</th>';
				h += '</tr></thead><tbody>';

				rows.forEach(function (row) {
					h += '<tr onclick="frappe.set_route(\'Form\',\'Stock Reservation\',\'' + row.reservation + '\')" style="cursor:pointer;" onmouseenter="this.style.background=\'#f8fafc\'" onmouseleave="this.style.background=\'transparent\'">';
					h += '<td style="padding:8px 12px;font-weight:700;color:#3b82f6;border-top:1px solid #f8fafc;">' + row.reservation + '</td>';
					h += '<td style="padding:8px 12px;border-top:1px solid #f8fafc;"><span style="font-weight:600;color:#1e293b;">' + (row.company || '').replace(' Enterprise', '').replace(' Export', '') + '</span></td>';
					h += '<td style="padding:8px 12px;font-size:10px;color:#64748b;border-top:1px solid #f8fafc;">' + (row.warehouse || '') + '</td>';
					h += '<td style="padding:8px 12px;font-weight:700;color:#1e293b;border-top:1px solid #f8fafc;">' + (row.item || '') + '</td>';
					h += '<td style="padding:8px 12px;text-align:right;font-weight:800;color:' + sc + ';border-top:1px solid #f8fafc;">' + frappe.utils.fmt_money(row.reserved_qty || 0) + '</td>';
					h += '<td style="padding:8px 12px;color:#64748b;border-top:1px solid #f8fafc;">' + (row.reserved_for || '') + '</td>';
					h += '<td style="padding:8px 12px;color:#64748b;border-top:1px solid #f8fafc;">' + (row.posting_date || '') + '</td>';
					h += '</tr>';
				});
				h += '</tbody></table></div>';
			});
		} else {
			h += '<div style="text-align:center;padding:40px;color:#94a3b8;">No reserved stock found</div>';
		}

		h += '</div>';
		frappe.query_report.page.main.find('.result').html(h);
	}
};
